akka-actor/src/main/scala/akka/dispatch/affinity/AffinityPool.scala

```diff
@@ -39,92 +125,85 @@ import scala.util.control.NonFatal
   extends AbstractExecutorService {
 
   if (parallelism <= 0)
     throw new IllegalArgumentException("Size of pool cannot be less or equal to 0")
 
+  import AffinityPool._
+
   // Held while starting/shutting down workers/pool in order to make
   // the operations linear and enforce atomicity. An example of that would be
   // adding a worker. We want the creation of the worker, addition
   // to the set and starting to worker to be an atomic action. Using
   // a concurrent set would not give us that
-  private val bookKeepingLock = new ReentrantLock()
+  private val bookKeepingLock = new ReentrantGuard()
 
   // condition used for awaiting termination
   private val terminationCondition = bookKeepingLock.newCondition()
 
   // indicates the current state of the pool
-  @volatile final private var poolState: PoolState = Running
+  @volatile final private var poolState: PoolState = Uninitialized
 
-  private final val workQueues = Array.fill(parallelism)(new BoundedTaskQueue(affinityGroupSize))
-  private final val workers = mutable.Set[ThreadPoolWorker]()
+  private final val workQueues = Array.fill(parallelism)(new BoundedAffinityTaskQueue(affinityGroupSize))
+  private final val workers = mutable.Set[AffinityPoolWorker]()
 
-  // a counter that gets incremented every time a task is queued
-  private val executionCounter: AtomicInteger = new AtomicInteger(0)
   // maps a runnable to an index of a worker queue
-  private val runnableToWorkerQueueIndex = new AtomicReference(ImmutableIntMap.empty)
-
-  private def locked[T](l: Lock)(body: ⇒ T) = {
-    l.lock()
-    try {
-      body
-    } finally {
-      l.unlock()
-    }
-  }
-
-  private def getQueueForRunnable(command: Runnable) = {
+  private[this] final val hashCache = new AtomicReference(ImmutableIntMap.empty)
 
+  private def getQueueForRunnable(command: Runnable): BoundedAffinityTaskQueue = {
     val runnableHash = command.hashCode()
 
-    def sbhash(i: Int) = reverseBytes(i * 0x9e3775cd) * 0x9e3775cd
-
-    def getNext = executionCounter.incrementAndGet() % parallelism
+    def indexFor(h: Int): Int =
+      Math.abs(reverseBytes(h * 0x9e3775cd) * 0x9e3775cd) % parallelism // In memory of Phil Bagwell
 
-    def updateIfAbsentAndGetQueueIndex(
-      workerQueueIndex: AtomicReference[ImmutableIntMap],
-      runnableHash:     Int, queueIndex: ⇒ Int): Int = {
-      @tailrec
-      def updateIndex(): Unit = {
-        val prev = workerQueueIndex.get()
-        if (!runnableToWorkerQueueIndex.compareAndSet(prev, prev.updateIfAbsent(runnableHash, queueIndex))) {
-          updateIndex()
+    val workQueueIndex =
+      if (fairDistributionThreshold == 0)
+        indexFor(runnableHash)
+      else {
+        @tailrec
+        def cacheLookup(prev: ImmutableIntMap, hash: Int): Int = {
+          val existingIndex = prev.get(runnableHash)
+          if (existingIndex >= 0) existingIndex
+          else if (prev.size > fairDistributionThreshold) indexFor(hash)
```

Comment:
@raboof It's rather likely that the sbhash improvments is much cheaper than the memory indirection of the cache. I considered adding a volatile field to AffinityPool and use s.m.U to access it as well as making the ImmutableIntMap methods static to operate on a plain Int[] which would mean no indirection through the AR and no indirection through the ImmutableIntMap wrapper type. If this was transparently hidden in a method to compute the queue to use then we could make the strategy configurable and self-contained (while still fast).